worker/uniter/relation/state.go

```diff
@@ -108,200 +107,196 @@ func (s *State) Validate(hi hook.Info) (err error) {
+	}
+	ops := NewStateOps(relationId, rw)
+	if err = yaml.Unmarshal([]byte(stateYaml), &ops.state); err != nil {
+		return nil, errors.Annotatef(err, "cannot unmarshall relation %d state", relationId)
 	}
-	return d, nil
+	return ops, nil
 }
 
-// ReadAllStateDirs loads and returns every StateDir persisted directly inside
-// the supplied dirPath. If dirPath does not exist, no error is returned.
-func ReadAllStateDirs(dirPath string) (dirs map[int]*StateDir, err error) {
-	defer errors.DeferredAnnotatef(&err, "cannot load relations state from %q", dirPath)
-	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
-		return nil, nil
-	} else if err != nil {
-		return nil, err
-	}
-	fis, err := ioutil.ReadDir(dirPath)
-	if err != nil {
-		return nil, err
+// ReadAllStates loads the State and returns a StateOps for every
+// persisted relation.
+func ReadAllStates(rw UnitStateReadWriter) (map[int]*StateOps, error) {
+	ops := map[int]*StateOps{}
+
+	unitState, err := rw.State()
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, errors.Trace(err)
 	}
-	dirs = map[int]*StateDir{}
-	for _, fi := range fis {
-		// Entries with integer names must be directories containing StateDir
-		// data; all other names will be ignored.
-		relationId, err := strconv.Atoi(fi.Name())
-		if err != nil {
-			// This doesn't look like a relation.
-			continue
+	for k, v := range unitState.RelationState {
+		var state State
+		if err = yaml.Unmarshal([]byte(v), &state); err != nil {
+			return nil, errors.Annotatef(err, "cannot unmarshall relation %d state", k)
 		}
-		dir, err := ReadStateDir(dirPath, relationId)
-		if err != nil {
-			return nil, err
+		ops[k] = &StateOps{
+			unitStateRW: rw,
+			state:       state,
 		}
-		dirs[relationId] = dir
 	}
-	return dirs, nil
-}
 
-// Ensure creates the directory if it does not already exist.
-func (d *StateDir) Ensure() error {
-	return os.MkdirAll(d.path, 0755)
+	return ops, nil
 }
 
-// Exists returns true if the directory for this state exists.
-func (d *StateDir) Exists() bool {
-	_, err := os.Stat(d.path)
-	return err == nil
-}
-
-// Write atomically writes to disk the relation state change in hi.
+// WriteHook atomically writes to disk the relation state change in hi.
 // It must be called after the respective hook was executed successfully.
-// Write doesn't validate hi but guarantees that successive writes of
+// WriteHook doesn't validate hi but guarantees that successive writes of
 // the same hi are idempotent.
-func (d *StateDir) Write(hi hook.Info) (err error) {
-	defer errors.DeferredAnnotatef(&err, "failed to write %q hook info for %q on state directory", hi.Kind, hi.RemoteUnit)
+func (s *StateOps) WriteHook(hi hook.Info) (err error) {
+	defer errors.DeferredAnnotatef(&err, "failed to write %q hook info for %q in state", hi.Kind, hi.RemoteUnit)
 	if hi.Kind == hooks.RelationBroken {
-		return d.Remove()
+		return s.Remove()
 	}
-	name := strings.Replace(hi.RemoteUnit, "/", "-", 1)
 	isApp := false
```

Comment:
Due to the removal, this block can be condensed to `isApp := hi.RemoteUnit == ""`

Reply:
done